src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs

```diff
@@ -0,0 +1,362 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core.Functions
+{
+    internal class TexlFunctionSet
+    {
+        // Dictionary key: function.Name
+        private Dictionary<string, List<TexlFunction>> _functions;
+
+        // Dictionary key: function.LocaleInvariantName
+        private Dictionary<string, List<TexlFunction>> _functionsInvariant;
+
+        // Dictionary key: function.Namespace
+        private Dictionary<DPath, List<TexlFunction>> _namespaces;
+
+        // List of all function.GetRequiredEnumNames()
+        private List<string> _enums;
+
+        // Count of functions
+        internal int _count;
+
+        internal Dictionary<string, List<TexlFunction>>.KeyCollection FunctionNames => _functions.Keys;
+
+        internal Dictionary<string, List<TexlFunction>>.KeyCollection InvariantFunctionNames => _functionsInvariant.Keys;
+
+        internal Dictionary<DPath, List<TexlFunction>>.KeyCollection Namespaces => _namespaces.Keys;
+
+        internal IEnumerable<TexlFunction> WithName(string name) => WithNameInternal(name);
+
+        private List<TexlFunction> WithNameInternal(string name) => _functions.TryGetValue(name, out List<TexlFunction> result) ? result : new List<TexlFunction>();
+
+        internal IEnumerable<TexlFunction> WithName(string name, DPath ns) => _functions.TryGetValue(name, out List<TexlFunction> result) ? result.Where(f => f.Namespace == ns) : new List<TexlFunction>();
+
+        internal IEnumerable<TexlFunction> WithInvariantName(string name) => WithInvariantNameInternal(name);
+
+        private List<TexlFunction> WithInvariantNameInternal(string name) => _functionsInvariant.TryGetValue(name, out List<TexlFunction> result) ? result : new List<TexlFunction>();
+
+        internal IEnumerable<TexlFunction> WithInvariantName(string name, DPath ns) => _functionsInvariant.TryGetValue(name, out List<TexlFunction> result) ? result.Where(f => f.Namespace == ns) : new List<TexlFunction>();
+
+        internal IEnumerable<TexlFunction> WithNamespace(DPath ns) => WithNamespaceInternal(ns);
+
+        private List<TexlFunction> WithNamespaceInternal(DPath ns) => _namespaces.TryGetValue(ns, out List<TexlFunction> result) ? result : new List<TexlFunction>();
+
+        internal IEnumerable<string> Enums => _enums;
+
+        internal TexlFunctionSet()
+        {
+            _functions = new Dictionary<string, List<TexlFunction>>();
```

Comment:
>new Dictionary<string, List<TexlFunction>>(); [](http://example.com/codeflow?start=25&length=45)

nit: needs string comparator

Reply:
I wouldn't say "nit". You're right, this is missing and should probably be made mandatory via a .Net analyzer if it exists